url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/SimpleThreadPool.java

```diff
@@ -0,0 +1,567 @@
+        }
+      }
+
+      return availWorkers.size();
+    }
+  }
+
+  private void makeAvailable(WorkerThread wt) {
+
+    synchronized (nextRunnableLock) {
+      if (!isShutdown) {
+        availWorkers.add(wt);
+      }
+      busyWorkers.remove(wt);
+      nextRunnableLock.notifyAll();
+    }
+  }
+
+  private void clearFromBusyWorkersList(WorkerThread wt) {
+
+    synchronized (nextRunnableLock) {
+      busyWorkers.remove(wt);
+      nextRunnableLock.notifyAll();
+    }
+  }
+
+  /*
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ WorkerThread Class.
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+
+  /**
+   * <p>
+   * A Worker loops, waiting to execute tasks.
+   * </p>
+   */
+  private class WorkerThread extends Thread {
+
+    // A flag that signals the WorkerThread to terminate.
+    private boolean run = true;
+
+    private SimpleThreadPool tp;
+
+    private Runnable runnable = null;
+
+    private boolean runOnce = false;
+
+    /**
+     * <p>
+     * Create a worker thread and start it. Waiting for the next Runnable, executing it, and waiting for the next Runnable, until the shutdown flag is
+     * set.
+     * </p>
+     */
+    private WorkerThread(SimpleThreadPool tp, ThreadGroup threadGroup, String name, int prio, boolean isDaemon) {
+
+      this(tp, threadGroup, name, prio, isDaemon, null);
+    }
+
+    /**
+     * <p>
+     * Create a worker thread, start it, execute the runnable and terminate the thread (one time execution).
+     * </p>
+     */
+    private WorkerThread(SimpleThreadPool tp, ThreadGroup threadGroup, String name, int prio, boolean isDaemon, Runnable runnable) {
+
+      super(threadGroup, name);
+      this.tp = tp;
+      this.runnable = runnable;
+      if (runnable != null) {
+        runOnce = true;
+      }
+      setPriority(prio);
+      setDaemon(isDaemon);
+    }
+
+    /**
+     * <p>
+     * Signal the thread that it should terminate.
+     * </p>
+     */
+    void shutdown() {
+
+      synchronized (this) {
+        run = false;
+      }
+    }
+
+    public void run(Runnable newRunnable) {
+
+      synchronized (this) {
+        if (runnable != null) {
+          throw new IllegalStateException("Already running a Runnable!");
+        }
+
+        runnable = newRunnable;
+        this.notifyAll();
+      }
+    }
+
+    /**
+     * <p>
+     * Loop, executing targets as they are received.
+     * </p>
+     */
+    @Override
+    public void run() {
+
+      boolean ran = false;
+      boolean shouldRun = false;
+      synchronized (this) {
+        shouldRun = run;
+      }
+
+      while (shouldRun) {
+        try {
+          synchronized (this) {
+            while (runnable == null && run) {
+              this.wait(500);
+            }
+
+            if (runnable != null) {
+              ran = true;
+              runnable.run();
+            }
+          }
+        } catch (InterruptedException unblock) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)